### PR TITLE
[ONEDNN][BC-breaking] update onednn from v2.7.3 to v3.1.1

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
**Summary**
Update onednn from v2.7.3 to v3.1.1.
It is bc-breaking as some APIs are changed on oneDNN side. Changes include:
- PyTorch code where oneDNN is directly called
- Submodule `third_party/ideep` to adapt to oneDNN's new API.
- CMAKE files to fix build issues.

**Test plan**
Building issues and correctness are covered by CI checks.
For performance, we have run TorchBench models to ensure there is no regression. Below is the comparison before and after oneDNN update. 
image

Note: 
- Base commit of PyTorch: da322ea
- CPU: Intel(R) Xeon(R) Platinum 8380 CPU @ 2.30GHz (Ice Lake)

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen

### CODE REVIEW

File: aten/src/ATen/native/Activation.cpp [hunk 1]
```diff
@@ -394,7 +394,8 @@ TORCH_IMPL_FUNC(gelu_out_cpu) (
 ) {
 auto approximate_type = get_gelutype_enum(approximate);
 #if AT_MKLDNN_ENABLED()
-  if (use_mkldnn(self) && (approximate_type == GeluType::None)) {
+  // TODO: oneDNN's gelu fails to provide enough accuracy. Disable for now.
+  if (0 && use_mkldnn(self) && (approximate_type == GeluType::None)) {
```
Comment (jgong5) [inline aten/src/ATen/native/Activation.cpp:398]: Do we expect this will be fixed in v3.1?

Comment (Xia-Weiwen) [reply aten/src/ATen/native/Activation.cpp:398]: Yes. The gelu issue should be fixed by oneDNN. I disabled mkldnn gelu here just to run CI checks. I will undo this once oneDNN's fix is available publicly.

File: third_party/mkl-dnn.BUILD [hunk 2]
```diff
@@ -122,4 +126,7 @@ cc_library(
         "@pytorch//tools/config:thread_sanitizer": [],
         "//conditions:default": ["@tbb"],
     }),
+    defines = [
```
Comment (Xia-Weiwen) [inline third_party/mkl-dnn.BUILD:129]: Hi @jgong5 We found that bazel test fails if we do not define this macro. We also found that the failure can be fixed alternatively by adding `#cmakedefine01 DNNL_ENABLE_MAX_CPU_ISA` to file `oneDNN/include/oneapi/dnnl/dnnl_config.h.in` and adding a rule like `"#cmakedefine01 DNNL_ENABLE_MAX_CPU_ISA": "#define DNNL_ENABLE_MAX_CPU_ISA1"` in this file.
Which do you think is better?

File: third_party/mkl-dnn.BUILD [hunk 3]
```diff
@@ -49,47 +50,50 @@ _DNNL_RUNTIME_OMP = {
...
+        "src/cpu/**/*.h",
+        "src/common/*.hpp",
+        "src/common/ittnotify/jitprofiling.h",
     ], exclude=[
-        "third_party/oneDNN/src/cpu/aarch64/**/*.hpp",
-        "third_party/oneDNN/src/cpu/aarch64/**/*.h",
+        "src/cpu/aarch64/**/*.hpp",
+        "src/cpu/aarch64/**/*.h",
+        "src/cpu/rv64/**/*.hpp",
+        "src/cpu/rv64/**/*.h",
```
Comment (Xia-Weiwen) [inline third_party/mkl-dnn.BUILD:93]: Hi @jgong5 We excluded RV64 files due to CI failures (build issue). Do you think it's OK?

Comment (jgong5) [reply third_party/mkl-dnn.BUILD:93]: Let's ignore these newly added RISCV files.

File: third_party/mkl-dnn.BUILD (hunk 2 above)
Comment (jgong5) [reply third_party/mkl-dnn.BUILD:129]: Let's add the definition here. Please also check and add all the default on macros here.

File: third_party/mkl-dnn.BUILD (hunk 3 above)
Comment (Xia-Weiwen) [reply third_party/mkl-dnn.BUILD:93]: OK

File: third_party/mkl-dnn.BUILD (hunk 2 above)
Comment (Xia-Weiwen) [reply third_party/mkl-dnn.BUILD:129]: OK. I have added other macro definitions here as well.

**Review by jgong5 — APPROVED** [review]

**Review by jerryzh168 — APPROVED** [review]

**Comment by Guobing-Chen** [comment]
@malfet, from the comment above by facebook-github-bot, seems you've imported this PR for Meta internal, is everything OK, can we merge this PR now?

**Comment by Xia-Weiwen** [comment]
Hi @malfet. The conflict has been fixed. The only CI failure is due to some instability and unrelated to this PR. Could you kindly take a look and import it for internal test again? Thanks.

**Comment by Xia-Weiwen** [comment]
Hi @malfet. We hope to land this soon. What do we need to do? Could you please take a look? Thanks.

**Comment by wenzhe-nrv** [comment]
@Xia-Weiwen since oneDNN 3.2 has been release, do you think this PR could upgrade to 3.2?

**Review by albanD — COMMENTED** [review]
Hi,

Could you clarify what exactly is BC-breaking about this PR? It is not clear to me from the code nor the PR description.

**Comment by vpirogov** [comment]
@albanD, you can find full list of oneDNN v3.x breaking changes in [the release notes](https://github.com/oneapi-src/oneDNN/releases/tag/v3.0). Pytorch is impacted by [quantization related changes described here](https://github.com/oneapi-src/oneDNN/blob/rfcs/rfcs/20220201-quantization-scaling/README.md).

**Comment by albanD** [comment]
So these are BC-breaking from the ondnn point of view but not from the PyTorch point of view? None of the functionality that PyTorch offers is changing in this PR. Or I'm missing some details here?

**Comment by malfet** [comment]
@albanD yes, there are no BC breaking changes from PyTorch user PoV, but it's quite painful to import it to internal build system due to ideep/onednn build changes

File: test/test_nn.py [hunk 4]
```diff
@@ -3295,6 +3295,7 @@ def perm_fn(x):
             self.assertEqual(tuple(result.shape), tuple(ref_output.shape))
             np.testing.assert_allclose(result, ref_output, atol=1e-5)
 
+    @set_default_dtype(torch.double)
```
Comment (kit1980) [inline test/test_nn.py:3298]: Why is this here?
Bad merge after https://github.com/pytorch/pytorch/pull/107069 ?
cc @kurtamohler

Comment (Guobing-Chen) [reply test/test_nn.py:3298]: The PR #107069 missed to add the decorator for this test case which makes it run with float type while based on the tolerance defined in the test case it should be double. Same decorator have been added for lots of other test cases in the same test file with the PR, while this one seems missed.

**Comment by Xia-Weiwen** [comment]
> @Xia-Weiwen since oneDNN 3.2 has been released, do you think this PR could upgrade to 3.2?

@wenzhe-nrv No. We only update oneDNN to 3.1.1 for now.

File: test/test_nn.py (hunk 4 above)
Comment (kurtamohler) [reply test/test_nn.py:3298]: Thanks for catching that!
